File: mobile/plugins/lola-transport/ios/Sources/LolaTransportCore/FrameCodec.swift
import Foundation

/// The wire framing, and nothing else.
///
/// Every frame on this transport is a four-byte big-endian `uint32` giving the
/// byte count of a UTF-8 JSON body, followed by exactly that many bytes. There
/// is no magic number, no type byte and no trailing newline; the frame's kind
/// lives inside the JSON. This mirrors `internal/protocol/framecodec.go` byte
/// for byte, and the golden vectors in `mobile/src/wire/testdata/frames.json`
/// are what hold the two honest.
///
/// The single most important property here is that an unacceptable length is
/// refused from the HEADER, before a byte of the body is waited for. A decoder
/// that accumulates first and checks afterwards will happily buffer toward a
/// four-gigabyte promise made by a peer, which is the whole reason the protocol
/// carries a prefix instead of a delimiter.
public enum LolaFrameLimits {
    /// Mirrors `protocol.FrameHeaderBytes`.
    public static let headerBytes = 4

    /// Mirrors `protocol.MaxFrameBytes` (1 MiB). It bounds the BODY, not the
    /// framed size, in both directions.
    public static let maxFrameBytes = 1 << 20
}

public enum FrameCodecError: Error, Equatable {
    /// A zero length prefix. There is no empty frame in this protocol — an
    /// envelope always carries at least `v` and `type` — so a zero is a bug or
    /// a probe. Fatal: nothing after it can be trusted to be a frame boundary.
    case emptyFrame

    /// A length prefix over `maxFrameBytes`. Also fatal, and for a sharper
    /// reason: skipping the frame would mean reading the very bytes the
    /// decoder just refused to read, so the stream cannot be resynchronised.
    case frameTooLarge(size: Int, max: Int)

    /// An outbound body over `maxFrameBytes`. The caller's bug; nothing is
    /// written, because a truncated frame is worse than no frame.
    case bodyTooLarge(size: Int, max: Int)

    /// The decoder already failed fatally and was not reset.
    case decoderPoisoned

    /// A body that is not valid UTF-8. The daemon writes JSON produced by
    /// `encoding/json`, which is always valid UTF-8, so this is a corrupt
    /// stream rather than a protocol variation.
    case invalidUTF8
}

/// Encodes frame bodies onto the wire.
public enum FrameEncoder {
    /// Prefixes one body. Throws `bodyTooLarge` rather than truncating.
    public static func encode(body: Data) throws -> Data {
        try encode(bodies: [body])
    }

    /// Prefixes several bodies into one contiguous buffer.
    ///
    /// The daemon writes one frame per socket write so a four-byte header never
    /// lands in its own TCP segment next to Nagle. Concatenating several
    /// COMPLETE frames into one write preserves that property and improves on
    /// it: the reader still sees well-formed frames back to back, and the
    /// keystroke path pays one syscall instead of several.
    public static func encode(bodies: [Data]) throws -> Data {
        var total = 0
        for body in bodies {
            guard body.count <= LolaFrameLimits.maxFrameBytes else {
                throw FrameCodecError.bodyTooLarge(size: body.count, max: LolaFrameLimits.maxFrameBytes)
            }
            guard !body.isEmpty else {
                // A zero-length body would encode as a zero prefix, which the
                // peer treats as fatal. Refuse it here, where there is still a
                // caller to blame.
                throw FrameCodecError.emptyFrame
            }
            total += LolaFrameLimits.headerBytes + body.count
        }

        var out = Data(capacity: total)
        for body in bodies {
            let n = UInt32(body.count)
            out.append(UInt8(truncatingIfNeeded: n >> 24))
            out.append(UInt8(truncatingIfNeeded: n >> 16))
            out.append(UInt8(truncatingIfNeeded: n >> 8))
            out.append(UInt8(truncatingIfNeeded: n))
            out.append(body)
        }
        return out
    }

    /// Convenience for a body that is already a `String` of JSON.
    public static func encode(json: String) throws -> Data {
        guard let body = json.data(using: .utf8) else { throw FrameCodecError.invalidUTF8 }
        return try encode(body: body)
    }

    /// Convenience for several JSON strings.
    public static func encode(jsonBodies: [String]) throws -> Data {
        var bodies: [Data] = []
        bodies.reserveCapacity(jsonBodies.count)
        for s in jsonBodies {
            guard let d = s.data(using: .utf8) else { throw FrameCodecError.invalidUTF8 }
            bodies.append(d)
        }
        return try encode(bodies: bodies)
    }
}

/// Reassembles frames from a byte stream.
///
/// This is the half naive implementations get wrong. TCP delivers segments, not
/// messages: a single `receive` may hand over half a length prefix, two and a
/// half frames, or one byte. So the decoder buffers, and `next()` returns `nil`
/// to mean "not enough yet" rather than treating a short read as an error.
///
/// It is a value type with `mutating` methods and is NOT thread-safe. One
/// decoder per connection, touched only from that connection's serial queue.
public struct FrameDecoder {
    private var buffer: [UInt8] = []
    private var offset: Int = 0
    private var failed = false

    /// Bytes held but not yet delivered as a frame.
    public var buffered: Int { buffer.count - offset }

    /// True once a fatal framing error has been thrown. Every later `next()`
    /// throws `decoderPoisoned` until `reset()` is called, because a stream
    /// whose length prefix could not be honoured has no recoverable boundary.
    public var poisoned: Bool { failed }

    /// Compaction is amortised: bytes are dropped from the front only once
    /// enough of them have accumulated to be worth a copy, so a stream of small
    /// frames does not turn into a quadratic memmove.
    private static let compactThreshold = 64 * 1024

    public init() {}

    /// Appends bytes read from the socket.
    public mutating func push(_ chunk: Data) {
        guard !chunk.isEmpty else { return }
        buffer.append(contentsOf: chunk)
    }

    public mutating func push(_ chunk: [UInt8]) {
        guard !chunk.isEmpty else { return }
        buffer.append(contentsOf: chunk)
    }

    /// Returns the next complete frame body, or `nil` when more bytes are
    /// needed. Throws only on a FATAL framing violation, which the caller must
    /// answer by closing the connection.
    public mutating func next() throws -> Data? {
        if failed { throw FrameCodecError.decoderPoisoned }

        let available = buffer.count - offset
        if available < LolaFrameLimits.headerBytes {
            compact()
            return nil
        }

        let n =
            Int(buffer[offset]) << 24 | Int(buffer[offset + 1]) << 16 | Int(buffer[offset + 2]) << 8
            | Int(buffer[offset + 3])

        // Both checks happen on the HEADER alone. Neither waits for the body,
        // which is exactly the property a delimiter-based reader cannot have.
        if n == 0 {
            failed = true
            throw FrameCodecError.emptyFrame
        }
        if n > LolaFrameLimits.maxFrameBytes {
            failed = true
            throw FrameCodecError.frameTooLarge(size: n, max: LolaFrameLimits.maxFrameBytes)
        }

        if available - LolaFrameLimits.headerBytes < n {
            compact()
            return nil
        }

        let start = offset + LolaFrameLimits.headerBytes
        let body = Data(buffer[start..<(start + n)])
        offset = start + n
        compact()
        return body
    }

    /// Drains every complete frame currently buffered.
    ///
    /// On a fatal error the frames decoded BEFORE it are still handed to
    /// `sink`, then the error is thrown. The daemon does the same thing in the
    /// other direction — it answers a bad prefix with a best-effort refusal
    /// frame and then closes — so a client that discarded the good frames it
    /// already had would be throwing away the very message that explains the
    /// close.
    public mutating func drain(into sink: (Data) -> Void) throws {
        while let body = try next() {
            sink(body)
        }
    }

    /// Clears everything, including the poisoned flag. Only meaningful when the
    /// underlying socket has been replaced.
    public mutating func reset() {
        buffer.removeAll(keepingCapacity: false)
        offset = 0
        failed = false
    }

    private mutating func compact() {
        if offset == buffer.count {
            buffer.removeAll(keepingCapacity: true)
            offset = 0
            return
        }
        if offset >= Self.compactThreshold {
            buffer.removeFirst(offset)
            offset = 0
        }
    }
}

File: internal/protocol/framecodec.go
package protocol

import (
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"sync"
)

// The framing is a 4-byte big-endian length prefix followed by exactly that
// many bytes of JSON. This file is the only I/O in the package, and it is here
// rather than in internal/remote because both ends of the wire have to agree on
// it byte for byte and neither owns the other (see the note on Frame).
//
// Why a prefix rather than the newline-delimited JSON the unix socket uses.
// internal/daemon's handleConn reads with a bufio.Scanner capped at 1 MiB, and a
// scanner must BUFFER TO THE DELIMITER before it can know a message's size: an
// oversized frame is only detected after a megabyte has been read into a
// doubling buffer, and Scanner then returns ErrTooLong with no way to
// resynchronize. With a prefix the reader takes 4 bytes, compares against
// MaxFrameBytes, and refuses BEFORE ALLOCATING. That does not matter much when
// the peer is a local CLI; it matters here, where the peer is a device on a
// network. The prefix also makes each frame self-delimiting on the wire, so a
// short write is a detectable desync rather than two silently merged lines —
// which is what makes it safe for several goroutines to interleave replies and
// pane output onto one connection, keyed by Frame.ID.
//
// The prefix is the STREAM ADAPTER and it disappears over a message transport:
// on a WebSocket each frame is one binary message and the transport already
// frames it. Keeping the JSON body identical either way is what lets the client
// codec be written once.

var (
	// ErrFrameTooLarge is a length prefix (inbound) or an encoded body
	// (outbound) exceeding MaxFrameBytes. Inbound it is FATAL to the
	// connection: skipping the frame would mean reading the very bytes the
	// reader just refused to read, and a stream whose length prefix cannot be
	// trusted cannot be resynchronized. Write the refusal best-effort and close.
	// Outbound it is a bug in the daemon: log and DROP the frame, never
	// truncate it — a truncated resync renders a wrong screen, which is worse
	// than no screen at all.
	ErrFrameTooLarge = errors.New("protocol: frame exceeds the maximum frame size")

	// ErrFrameEmpty is a zero-length prefix. There is no empty frame in this
	// protocol — the envelope always carries at least V and Type — so a zero is
	// either a bug or a probe. Fatal to the connection, same reasoning as
	// ErrFrameTooLarge: nothing downstream can be trusted to be a frame boundary.
	ErrFrameEmpty = errors.New("protocol: zero-length frame")

	// ErrFrameMalformed is a body that is not a decodable envelope. The framing
	// itself was intact, so the connection MAY survive when the caller can tell
	// which frame it was (a bad payload on a good envelope is answered on that
	// frame's ID); a malformed ENVELOPE has no id to answer on and is closed.
	// The wrapped text is for the local log only: it must never be echoed to the
	// peer, which gets CodeInternal and a fixed message.
	ErrFrameMalformed = errors.New("protocol: malformed frame")

	errNilFrame = errors.New("protocol: ReadFrame into a nil Frame")
)

// FrameReader reads length-prefixed frames from a stream. It is NOT safe for
// concurrent use: one reader goroutine per connection, which is the shape the
// correlation id exists to make sufficient.
//
// It keeps ONE body buffer and reuses it across calls, so after the first frame
// of a given size the hot path allocates nothing for the body itself. The buffer
// grows only to the largest frame actually seen and can never exceed
// MaxFrameBytes, because the size is checked against that cap before a single
// byte is allocated.
type FrameReader struct {
	r   io.Reader
	hdr [FrameHeaderBytes]byte
	buf []byte
}

// NewFrameReader wraps r. Callers on a network connection are expected to set a
// read deadline on the connection itself and refresh it per frame; the codec
// deliberately owns no timeouts, because the right deadline differs between a
// request stream and an attached pane.
func NewFrameReader(r io.Reader) *FrameReader { return &FrameReader{r: r} }

// ReadFrame decodes the next frame into dst.
//
// dst is RESET FIRST — every field zeroed, the Payload's backing array kept for
// reuse — because encoding/json leaves struct fields the JSON does not mention
// untouched. Reusing a Frame without that reset is how a previous frame's Cmd or
// Pane silently becomes this frame's, which is a fail-OPEN authorization bug
// rather than a cosmetic one.
//
// The returned Payload aliases dst's own buffer and is therefore only valid
// until the next ReadFrame into the SAME dst: a caller that hands the payload to
// another goroutine must copy it, or decode it before reading on.
//
// Errors: io.EOF exactly when the peer closed cleanly on a frame boundary;
// io.ErrUnexpectedEOF for a truncated header or body; ErrFrameEmpty and
// ErrFrameTooLarge for a length prefix that cannot be honoured (both fatal to
// the connection); ErrFrameMalformed for a body that is not an envelope. Nothing
// here validates V or Type — that is dispatch policy, and it lives with the
// authorizer that also needs the frame's identity. Use SupportedFrameVersion and
// DaemonAcceptsFrame / ClientAcceptsFrame for it.
func (fr *FrameReader) ReadFrame(dst *Frame) error {
	if dst == nil {
		return errNilFrame
	}
	if _, err := io.ReadFull(fr.r, fr.hdr[:]); err != nil {
		// io.ReadFull reports a clean close as io.EOF and a partial header as
		// io.ErrUnexpectedEOF; both are passed through unchanged so the caller
		// can tell "peer went away" from "peer went away mid-frame".
		return err
	}
	n := binary.BigEndian.Uint32(fr.hdr[:])
	if n == 0 {
		return ErrFrameEmpty
	}
	if n > MaxFrameBytes {
		return fmt.Errorf("%w: %d bytes (max %d)", ErrFrameTooLarge, n, MaxFrameBytes)
	}
	if cap(fr.buf) < int(n) {
		fr.buf = make([]byte, n)
	}
	body := fr.buf[:n]
	if _, err := io.ReadFull(fr.r, body); err != nil {
		if errors.Is(err, io.EOF) {
			// A prefix promised n bytes and the stream ended short. That is a
			// truncated frame, never a clean close.
			return io.ErrUnexpectedEOF
		}
		return err
	}
	*dst = Frame{Payload: dst.Payload[:0]}
	if err := json.Unmarshal(body, dst); err != nil {
		return fmt.Errorf("%w: %w", ErrFrameMalformed, err)
	}
	return nil
}

// FrameWriter writes length-prefixed frames to a stream. Unlike FrameReader it
// IS safe for concurrent use, and it has to be: replies, pane output and
// resyncs are produced by different goroutines and share one connection.
type FrameWriter struct {
	w  io.Writer
	mu sync.Mutex
}

// NewFrameWriter wraps w.
func NewFrameWriter(w io.Writer) *FrameWriter { return &FrameWriter{w: w} }

// WriteFrame encodes f and writes it as one length-prefixed message.
//
// The body is marshalled OUTSIDE the lock, so a slow encode never blocks the
// other goroutines sharing this connection, and the prefix is copied in front of
// it so the whole frame reaches the transport in ONE Write — a separate 4-byte
// write would put a header in its own TCP segment and interact badly with Nagle
// on the exact stream that carries keystrokes. The cost is one extra copy per
// frame, which at a coalesced pane's frame rate is not measurable; the
// allocation that actually mattered was the reader's doubling buffer, and the
// length prefix is what removes it.
//
// A body over MaxFrameBytes returns ErrFrameTooLarge and writes NOTHING.
// Truncating is never an option: half a resync paints a wrong screen.
func (fw *FrameWriter) WriteFrame(f *Frame) error {
	body, err := json.Marshal(f)
	if err != nil {
		return fmt.Errorf("protocol: encode frame: %w", err)
	}
	if len(body) > MaxFrameBytes {
		return fmt.Errorf("%w: %d bytes (max %d)", ErrFrameTooLarge, len(body), MaxFrameBytes)
	}
	out := make([]byte, FrameHeaderBytes+len(body))
	binary.BigEndian.PutUint32(out, uint32(len(body)))
	copy(out[FrameHeaderBytes:], body)

	fw.mu.Lock()
	defer fw.mu.Unlock()
	_, err = fw.w.Write(out)
	return err
}
